[New Payment Gateway] Minimal Payment Gateway Integration

pikender opened the pull request:
Use it for Reference

Commit:
[New Payment Gateway] Minimal Payment Gateway Integration
and it works :)

Commit:
[New Payment Gateway] Reference New Payment Gateway Config

nimish-mehta commented on apps/ex_shop/web/gateways/gateway.ex:
```diff
@@ -18,6 +18,10 @@ defmodule ExShop.Gateway do
   end
 
+  defp do_authorize_payment(order, "new_payment_gateway", payment_method_params) do
+    ExShop.Billing.Gateway.NewPaymentGateway.authorize(order, payment_method_params["new_payment_gateway"])
```
Here we are invoking the payment gateway directly, We start a dedicated worker thread for each payment gateway when the application starts and the gateway can be consumed via a common interface. See: https://github.com/vinsol/ex-shop/blob/umbrella/apps/ex_shop/web/gateways/stripe.ex#L12 for example.

pikender commented on apps/ex_shop/lib/ex_shop.ex:
```diff
@@ -15,7 +15,8 @@ defmodule ExShop do
       worker(Commerce.Billing.Worker, stripe_worker_configuration, id: :stripe),
-      worker(Commerce.Billing.Worker, braintree_worker_configuration, id: :braintree)
+      worker(Commerce.Billing.Worker, braintree_worker_configuration, id: :braintree),
+      worker(Commerce.Billing.Worker, new_payment_gateway_configuration, id: :new_payment_gateway)
```
Id should be different to start same worker with different configuration.

Remember, id is optional parameter so can be skipped but workers won't start without it :)

Commit:
Use Commerce.Billing.authorize(order, payment_method_params["new_payment_gateway"]) instead of
ExShop.Billing.Gateway.NewPaymentGateway.authorize(order, payment_method_params["new_payment_gateway"])

Use Commerce.Billing instead of Module implementing authorise
to leverage GenServer created for each payment Gateway
Below, will dispatch authorise to the dedicated GenServer for it
which in turn will call App Module implementing authorise
Direct call like above will leave the GenServer created useless / idle
Refer web/gateways/braintree_impl.ex on example to extract Billing Address
